accepted/00000-full-stack-type-safety.md

```diff
@@ -0,0 +1,102 @@
+- Feature Name: Full Stack Type Safety
+- Start Date: 2021-05-28
+
+# Summary
+[summary]: #summary
+
+Add type safety to frontend and backend communications to achieve type safety across the whole project.
+
+# Motivation
+[motivation]: #motivation
+
+With the wrapup of the Typescript migration, we now have type safety both in the backend project by itself and in the frontend project by itself. What we do not have is type safety in the communications between the two. If an entity changes on the backend but the corresponding code is not updated on the frontend, the regression can go unnoticed unless this specific interaction is covered by an E2E test. Adding type checks to backend and frontend communications will catch these regressions. In addition, shared types will also help write sturdier frontend code since it will alert the developer to possible nullable values and other similar edge cases which might otherwise go unnoticed.  
+
+# Detailed design
+[design]: #detailed-design
+
+In the order of increased difficulty but also increased payoff, there's two possible approaches. The easier but more laborious option is to generate a set of exposed types and manually annotate requests.  
+
+```ts
+// Type definitions generated from the backend project...
+namespace Space {
+  export interface Foo {
+    goodExample: boolean;
+  }
+}
+```
+
+```ts
+// ...can be used to manually annotate expected types in the frontend project.
+- const result = await Network.get("/rhn/manager/getFoo");
++ const result = await Network.get<Space.Foo>("/rhn/manager/getFoo");
+
+// 'goodExample' is correctly identified as a boolean value
+if (result.goodExample) {
+  // ...
+}
+
+// Throws an error whereas previously this passed silently
+if (result.badExample) {
+  //       ^^^^^^^^^^
+  // Property 'badExample' does not exist on type 'Foo'
+}
+```
+
+The general workflow for using this approach would be to add a step to the backend build process which generates a type declaration file as shown above. There are existing tools which provide this functionality, I've considered evaluating them out of scope for the time being. The generated file can be added to the frontend project scope and the type definitions can be used accordingly where needed. This change can be adopted gradually over time.  
```

Comment:
I am not against the overall idea, but think it's critical to define how this type declaration file can be generated from the backend code.

Without that piece, I guess the only possibility would be to create it manually. But in my understanding, that would only move the problem one step away, from "backend to frontend" type coherence to "backend to declararation-file" type coherence - correct?

Reply:
That is correct, yes. Without automated tooling, I don't think this proposal is reasonable. I haven't had the chance to see this approach in action on a Java project before so I don't know what tools may be applicable or realistically pluggable with our architecture. At my previous job we had a good run using [NSwag](https://github.com/RicoSuter/NSwag) on a .NET Core project, the flow there was that our backend project had Swagger tacked on and NSwag used the JSON output by Swagger to generate Typescript types.  

NSwag did have one major drawback in that the generated type names were not consistently unique. You could have entities with the same name in different business scopes, a-la both named `Entity`, mapped to names such as `Entity1` and `Entity2` in the output but when `Entity1` was removed in a refactor, `Entity2` would then become just `Entity`. That was a big pain point for us, I haven't checked whether they've found a fix for it since.